### PR TITLE
Refactor StreamD/IsStream for terminating folds

### PR DESCRIPTION
**Pull request by harendra-kumar**
* IsStream:
  * Add foldMany
  * Update haddock docs, with fold equivalents. Now with terminating folds the
    fold behvaior is equivalent to these folds.
  * Add skeletons for foldSequence/foldIterate/parseSequence
  * Implement combinators directly in terms of foldMany/foldMany1

StreamD:
  * export foldMany/foldMany1
  * Remove groupsOf/splitBy/splitSuffixBy/splitSuffixWith, we can use
    foldMany/foldMany1 directly instead.
  * Make some stylistic changes to code

StreamD/Types:
  * Remove groupsOf
  * Remove GroupConsume state from foldMany/foldMany1

* Add benchmarks for foldMany

### CODE REVIEW

File: src/Streamly/Internal/Data/Stream/StreamD/Type.hs [hunk 1]
```diff
@@ -607,15 +604,14 @@ foldMany (Fold fstep initial extract) (Stream step state) =
...
-            FL.Partial ps -> return $ Skip (GroupBuffer st ps)
     step' gst (GroupBuffer st fs) = do
         r <- step (adaptState gst) st
         case r of
-            Yield x s -> return $ Skip $ GroupConsume s fs x
+            Yield x s -> do
+                res <- fstep fs x
+                case res of
+                    FL.Done b -> return $ Skip (GroupYield b (GroupStart s))
+                    FL.Partial ps -> return $ Skip (GroupBuffer s ps)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Type.hs:614]:
```suggestion
                      FL.Done b -> return $ Skip (GroupYield b (GroupStart s))
                      FL.Partial ps -> return $ Skip (GroupBuffer s ps)
```

Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Type.hs:614]:
```suggestion
                case res of
                      FL.Done b -> return $ Skip (GroupYield b (GroupStart s))
                      FL.Partial ps -> return $ Skip (GroupBuffer s ps)
```

File: src/Streamly/Internal/Data/Stream/StreamD/Type.hs [hunk 2]
```diff
@@ -630,37 +626,33 @@ foldMany1 (Fold fstep initial extract) (Stream step state) =
...
     where
 
+    {-# INLINE consume #-}
+    consume x s fs = do
+        res <- fstep fs x
+        return
+            $ Skip
+            $ case res of
+                FL.Done b -> GroupYield b (GroupStart s)
+                FL.Partial ps -> GroupBuffer s ps
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Type.hs:636]:
```suggestion
            $ case res of
                  FL.Done b -> GroupYield b (GroupStart s)
                  FL.Partial ps -> GroupBuffer s ps
```

File: src/Streamly/Internal/Data/Stream/StreamD/Type.hs (hunk 1 above)
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Type.hs:614]:
```suggestion
                return
                    $ Skip
                    $ case res of
                           FL.Done b -> GroupYield b (GroupStart s)
                           FL.Partial ps -> GroupBuffer s ps
```

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 3]
```diff
@@ -4067,21 +4052,22 @@ tapOffsetEvery offset n (Fold fstep initial extract) (Stream step state) =
...
-                -- XXX Abstract the then branch?
-                then do
-                    acc1 <- fstep acc x
-                    return
-                        $ case acc1 of
-                              FL.Partial sres ->
-                                  Yield x $ TapOffTapping sres s (n - 1)
-                              FL.Done _ -> Yield x (TapOffDone s)
-                else return $ Yield x $ TapOffTapping acc s (count - 1)
+            Yield x s -> do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:4055]: I don't see any advantage in this refactor.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 4]
```diff
@@ -3767,10 +3755,8 @@ scanlMx' fstep begin done s =
 
 data PostScanState fsM s a = PostScan s !fsM
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:3756]:
```suggestion
data PostScanState s fsM = PostScan s !fsM
```

File: src/Streamly/Internal/Data/Stream/StreamD.hs (hunk 3 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD.hs:4055]: Just floats out the Yield, has lesser yield points, if that helps the compiler.

**Pull request closed without merging**